core/server/src/main/java/alluxio/master/file/meta/InodeTree.java

```diff
@@ -170,89 +187,251 @@ public boolean inodeIdExists(long id) {
+   * inode is not required to exist.
+   *
+   * @param path the path to lock
+   * @param lockMode the {@link LockMode} to lock the inodes with
+   * @return the {@link InodePath} representing the locked path of inodes
+   * @throws InvalidPathException if the path is invalid
    */
-  public Inode<?> getInodeById(long id) throws FileDoesNotExistException {
-    Inode<?> inode = mInodes.getFirstByField(mIdIndex, id);
-    if (inode == null) {
-      throw new FileDoesNotExistException("Inode id " + id + " does not exist.");
+  public InodePath lockInodePath(AlluxioURI path, LockMode lockMode) throws InvalidPathException {
+    TraversalResult traversalResult =
+        traverseToInode(PathUtils.getPathComponents(path.getPath()), lockMode, null);
+    return new ExtensibleInodePath(path, traversalResult.getInodes(),
+        traversalResult.getInodeLockGroup());
+  }
+
+  /**
+   * Locks existing inodes on the two specified paths. The two paths will be locked in the
+   * correct order. The target inodes are not required to exist.
+   *
+   * @param path1 the first path to lock
+   * @param lockMode1 the {@link LockMode} of the first path
+   * @param path2 the second path to lock
+   * @param lockMode2 the {@link LockMode} of the second path
+   * @return a {@link InodePathPair} representing the two locked paths
+   * @throws InvalidPathException if a path is invalid
+   */
+  public InodePathPair lockInodePathPair(AlluxioURI path1, LockMode lockMode1, AlluxioURI path2,
+      LockMode lockMode2) throws InvalidPathException {
+    String[] pathComponents1 = PathUtils.getPathComponents(path1.getPath());
+    String[] pathComponents2 = PathUtils.getPathComponents(path2.getPath());
+    List<LockMode> lockHints = new ArrayList<>();
+
+    int minLength = Math.min(pathComponents1.length, pathComponents2.length);
+    for (int i = 0; i < minLength; i++) {
+      if (pathComponents1[i].equals(pathComponents2[i])) {
+        // The two paths share a common path prefix.
+        LockMode mode1 = getLockModeForComponent(i, pathComponents1.length, lockMode1, null);
+        LockMode mode2 = getLockModeForComponent(i, pathComponents2.length, lockMode1, null);
+        // If either of the modes are WRITE, lock both paths as WRITE to prevent deadlock.
+        if (mode1 == LockMode.READ && mode2 == LockMode.READ) {
+          lockHints.add(LockMode.READ);
+        } else {
+          lockHints.add(LockMode.WRITE);
+        }
+      } else {
+        // The two paths no longer share a common prefix.
+        break;
+      }
+    }
+
+    // Lock paths in a deterministic order.
+    boolean switched = false;
+    AlluxioURI uriFirst = path1;
+    LockMode lockModeFirst = lockMode1;
+    String[] componentsFirst = pathComponents1;
+    AlluxioURI uriSecond = path2;
+    LockMode lockModeSecond = lockMode2;
+    String[] componentsSecond = pathComponents2;
+    if (path1.compareTo(path2) > 0) {
```

Comment:
I think you only need this check to determine in which order to call traverseToInodes, so you could do:

```
TraversalResult traversalResult1 = null;
TraversalResult traversalResult2 = null;
boolean valid = false;
try {
  if (path1.compareTo(path2) > 0) {
    traversalResult2 = traverseToInode(pathComponents2, lockMode2, lockHints);
    traversalResult1 = traverseToInode(pathComponents1, lockMode1, lockHints);
  } else {
    traversalResult1 = traverseToInode(pathComponents1, lockMode1, lockHints);
    traversalResult2 = traverseToInode(pathComponents2, lockMode2, lockHints);
  }

  InodePath inodePath1 = new ExtensibleInodePath(uriFirst, traversalResult1.getInodes(), traversalResult1.getInodeLockGroup());

  InodePath inodePath2 = new ExtensibleInodePath(uriSecond, traversalResult2.getInodes(),
traversalResult2.getInodeLockGroup());

  valid = true;
  return new InodePathPair(inodePath1, inodePath2);
} finally {
  if (!valid) {
    if (traversalResult1 != null) {
      traversalResult1.getInodeLockGroup().close();
    }
    if (traversalResult2 != null) {
      traversalResult2.getInodeLockGroup().close();
    }
 }
```